applications/serial_lte_modem/src/twi/slm_at_twi.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <drivers/i2c.h>
+#include <stdio.h>
+#include "slm_util.h"
+#include "slm_at_twi.h"
+
+LOG_MODULE_REGISTER(slm_twi, CONFIG_SLM_LOG_LEVEL);
+
+#if defined(CONFIG_HAS_HW_NRF_TWIM3)
+#define TWI_MAX_INSTANCE	4
+#elif defined(CONFIG_HAS_HW_NRF_TWIM2)
+#define TWI_MAX_INSTANCE	3
+#elif defined(CONFIG_HAS_HW_NRF_TWIM1)
+#define TWI_MAX_INSTANCE	2
+#elif defined(CONFIG_HAS_HW_NRF_TWIM0)
+#define TWI_MAX_INSTANCE	1
+#endif
+#define TWI_ADDR_LEN		2
+#define TWI_DATA_LEN		255
+
+#if (TWI_DATA_LEN * 2) > (CONFIG_SLM_SOCKET_RX_MAX * 2)
+# error "Please specify smaller TWI_DATA_LEN"
+#endif
+
+static const struct device *slm_twi_dev[TWI_MAX_INSTANCE];
+static uint8_t twi_data[TWI_DATA_LEN * 2 + 1];
+
+/* global functions defined in different files */
+void rsp_send(const uint8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+static void do_twi_list(void)
+{
+	memset(rsp_buf, 0, sizeof(rsp_buf));
+	sprintf(rsp_buf, "\r\n#XTWILS: ");
+
+	for (int i = 0; i < TWI_MAX_INSTANCE; i++) {
+		if (slm_twi_dev[i]) {
+			sprintf(rsp_buf + strlen(rsp_buf), "%d,", i);
+		}
+	}
+	if (strlen(rsp_buf) > 0) {
+		sprintf(rsp_buf + strlen(rsp_buf) - 1, "\r\n");
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	}
+}
+
+static int do_twi_write(uint16_t index, uint16_t dev_addr,
+			const uint8_t *twi_data_ascii, uint16_t ascii_len)
+{
+	int ret = -EINVAL;
+
+	if (!slm_twi_dev[index]) {
+		LOG_ERR("TWI device is not opened");
+		return ret;
+	}
+
+	/* Decode hex string to hex array */
+	memset(rsp_buf, 0, sizeof(rsp_buf));
+	ret = slm_util_atoh(twi_data_ascii, ascii_len, rsp_buf, ascii_len / 2);
+	if (ret < 0) {
+		LOG_ERR("Fail to decode hex string to hex array");
+		return ret;
+	}
+
+	ret = i2c_write(slm_twi_dev[index], rsp_buf, ret, dev_addr);
+	if (ret < 0) {
+		LOG_ERR("Fail to write twi data at address: %hx", dev_addr);
+	}
+
+	return ret;
+}
+
+static int do_twi_read(uint16_t index, uint16_t dev_addr, uint8_t num_read)
+{
+	int ret = -EINVAL;
+
+	if (!slm_twi_dev[index]) {
+		LOG_ERR("TWI device is not opened");
+		return ret;
+	}
+
+	if (num_read > TWI_DATA_LEN) {
+		LOG_ERR("Not enough buffer. Increase TWI_DATA_LEN");
+		return -ENOBUFS;
+	}
+
+	memset(twi_data, 0, sizeof(twi_data));
+	ret = i2c_read(slm_twi_dev[index], twi_data, (uint32_t)num_read, dev_addr);
+	if (ret < 0) {
+		LOG_ERR("Fail to read twi data");
+		return ret;
+	}
+	memset(rsp_buf, 0, sizeof(rsp_buf));
+	ret = slm_util_htoa(twi_data, num_read, rsp_buf, num_read * 2);
+	if (ret > 0) {
+		sprintf(rsp_buf + ret, "\r\n#XTWIR:\r\n");
```

Comment:
"\r\n" in the middle of URC could cause parsing issue

Reply:
I agree. Noe the \r\n before data is removed.